modules/client/src/client-common/modules/graphql.ts

```diff
@@ -4,40 +4,56 @@ import { Context } from "../../client-common/context";
 import { QueryStatus } from "../graphql-queries";
 import { IClientGraphQLCore } from "../interfaces/core";
 
-export class GraphqlModule implements IClientGraphQLCore {
-  private _graphql: GraphQLClient[] = [];
-  private _graphqlIdx: number = -1;
+const clientsMap = new Map<GraphqlModule, GraphQLClient[]>();
+const clientsIdxMap = new Map<GraphqlModule, number>();
 
+export class GraphqlModule implements IClientGraphQLCore {
   constructor(context: Context) {
+    // Storing client data in the private module's scope to prevent external mutation
     if (context.graphql?.length) {
-      this._graphql = context.graphql;
-      this._graphqlIdx = Math.floor(Random.getFloat() * context.graphql.length);
+      clientsMap.set(this, context.graphql);
+      clientsIdxMap.set(
+        this,
+        Math.floor(Random.getFloat() * context.graphql.length),
+      );
     }
+    Object.freeze(GraphqlModule.prototype);
+    Object.freeze(this);
   }
 
+  private get clients(): GraphQLClient[] {
+    return clientsMap.get(this) || [];
+  }
+  private get clientIdx(): number {
+    const idx = clientsIdxMap.get(this);
+    if (idx === undefined) {
+      return -1;
+    }
```

Comment:
Wouldn't it be easier to always set the `idx` to `-1` if there is no client in the constructor? This way, we wouldn't need to test for `undefined` every single time

Reply:
Same applies to the rest

Reply:
The problem with that is that if you use `Map.get()` it will return `number || undefined` and if you try to use an oneliner in this return like `return providerIdxMap.get(this) || -1` it wont work because if the map contains a 0 it will equal to `false`